pyoomph/__main__.py

```diff
@@ -101,97 +99,132 @@ def get_compilation_dir(parent_dir: pathlib.Path, name_hint: str) -> pathlib.Pat
+    compilation_units: List[CompilationUnit],
+    dependency_graph: Dict[Path, List[Path]],
+) -> List[CompilationUnit]:
     compilation_order: List[CompilationUnit] = []
-    while len(compilation_order) < len(dependencies):
-        unit = [u for u in all_compilation_units if u not in compilation_order][0]
-        decisions = [unit]
+    while len(compilation_order) < len(dependency_graph):
+        candidate_unit = next(
+            u for u in compilation_units if u not in compilation_order
+        )
+        breadcrumbs = [candidate_unit]
         while True:
-            need_first = [
+            uncompiled_dependencies = [
                 u
-                for u in all_compilation_units
+                for u in compilation_units
                 if u not in compilation_order
-                and u.source_path in dependencies[unit.source_path]
+                and u.source_path in dependency_graph[candidate_unit.source_path]
             ]
-            if not need_first:
+            if not uncompiled_dependencies:
                 break
-            if compiler_args.verbose:
+            if verbose:
                 print(
-                    f"Noticed dependency: {unit.source_path} --> {need_first[0].source_path}"
+                    f"Noticed dependency: {candidate_unit.source_path} --> {uncompiled_dependencies[0].source_path}"
+                )
+            candidate_unit = uncompiled_dependencies[0]
+            if candidate_unit in breadcrumbs:
+                message = (
+                    " --> ".join(d.source_path.name for d in breadcrumbs)
+                    + " --> "
+                    + candidate_unit.source_path.name
                 )
-            unit = need_first[0]
-            decisions.append(unit)
-            if decisions.count(unit) >= 2:
-                message = " --> ".join(d.source_path.name for d in decisions)
                 raise RuntimeError("cyclic imports: " + message)
-        compilation_order.append(unit)
+            breadcrumbs.append(candidate_unit)
+        compilation_order.append(candidate_unit)
+    return compilation_order
 
+
+def main() -> None:
+    arg_parser = argparse.ArgumentParser()
+    arg_parser.add_argument("infile", type=Path)
+    arg_parser.add_argument("-o", "--outfile", type=Path)
+    arg_parser.add_argument("--valgrind", default="")
+    arg_parser.add_argument("-v", "--verbose", action="store_true")
+    compiler_args, program_args = arg_parser.parse_known_args()
+
+    try:
+        cache_dir = compiler_args.infile.parent / ".oomph-cache"
+        cache_dir.mkdir(exist_ok=True)
+    except OSError:
+        cache_dir = Path.cwd() / ".oomph-cache"
+        cache_dir.mkdir(exist_ok=True)
+
+    # Create a compiler session
+    session = c_output.Session(
+        get_compilation_dir(cache_dir, compiler_args.infile.stem + "_compilation")
+    )
+
+    # Calculate the dependency graph
+    compilation_units, dependency_graph = compute_dependency_graph(
```

Comment:
Can we use `dependency_graph.keys()` instead of returning `compilation_units` separately?

Reply:
No, because `dependency_graph` contains paths. But can we change `dependency_graph` to contain compilation units instead of paths? There seems to already be code that looks up compilation units corresponding to what it found from `dependency_graph`.

Reply:
I fixed this myself. What do you think?

Reply:
I'm not sure that I'm a big fan, because it turns a O(1) lookup of a path into a O(n) lookup where you have to convert CUs into Paths. Perhaps we could change this by making CompilationUnits a dataclass that hash based on their path or something like that?

Reply:
From a practical point of view, I don't think `source_path in (...)` will ever be a performance bottleneck. The code already contains other `O(n)` lookups too, and `n` is quite small here, 10 or so when compiling the self-hosted compiler. Compare that to say number of generate C files, which is currently about 150 when compiling the self-hosted compiler